### PR TITLE
[ONNX] Support group arg of Convolution DKKC8 opt

### PR DESCRIPTION
**Pull request by beicy**

### CODE REVIEW

**Comment by beicy** [comment]
the following number is the avg by running 100 iterations.
resnet50 : 0.3678 (w group arg) vs. 0.3708 (w/o group arg)
vgg19      : 1.4078 (w group arg)  vs. 1.4167 (w/o group arg)
squeezenet: 0.0362 (w group arg) vs. 0.0361(w/o group arg)

File: tests/unittests/JITTest.cpp [hunk 1]
```diff
@@ -561,6 +561,78 @@ TEST_P(BackendCorrectnessTest, smallConv) {
   EXPECT_TRUE(out1.isEqual(out2));
 }
 
+TEST_P(CPUOnly, convDKKC8Test) {
```
Comment (jfix71) [inline tests/unittests/JITTest.cpp]: The test creates a Conv and assumes it will be optimized into a convDKKC8, but future things could change and that may no longer occur. Perhaps this should be called `groupConvTest` instead?

File: tests/unittests/JITTest.cpp [hunk 2]
```diff
@@ -561,6 +561,78 @@ TEST_P(BackendCorrectnessTest, smallConv) {
...
+  ConvolutionNode *CN =
+      F_->createConv("Conv", input, filter, zeroBias, outTy, 1, 1, 0, 2);
+  SaveNode *S = F_->createSave("save", CN);
+
+  EE_.compile(CompilationMode::Infer, F_);
+  EE_.run({}, {});
+
+  auto result = llvm::cast<Variable>(S->getOutput())->getPayload().getHandle();
+
+  std::array<float, 256> ref = {
```
Comment (jfix71) [inline tests/unittests/JITTest.cpp]: Why is this necessary? The results should be the same for the Interpreter's group conv, so you should be able to follow the pattern of the other tests here in JITTest -- run the same group conv on the Interpreter and the `backendKind_ ` and then `EXPECT_TRUE(out1.isEqual(out2));`.

Comment (beicy) [reply tests/unittests/JITTest.cpp]: I tried to follow the tests pattern in JITTest, passing the Tensors and the backendKind_. However, when I tried to get VarFrom(Tensor), the visibilityKind is public, which is necessary for the function ExecutionEngine::run() since there is an assert in ExecutionEngine::run  before laodValueFromTensor. However, for the DKKC8 opt, the filter is to be private. So I decided to add the ref result here and compare with it directly.

Comment (nadavrot) [reply tests/unittests/JITTest.cpp]: @beicy I am not sure I understand why we need this. The compiler can take a network such as Resnet and VGG and perform the optimization that changes the layout internally. Why can't you just create a small convolutional network and let the compiler optimize it?  Any JIT test that uses convolutional neural network should run the DKKC path.

File: lib/Backends/CPU/Transforms.cpp [hunk 3]
```diff
@@ -34,11 +34,12 @@ using llvm::isa;
...
   auto depth = CN->getFilter().dims()[0];
   auto *M = F->getParent();
+  auto group = CN->getGroup();
 
-  // The depth dimension must be a multiple of 64 to perform the
+  // The depth dimension in each group must be a multiple of 64 to perform the
   // transformation. This transformation is currently only profitable on
   // low-channel convolutions.
-  if (depth < 64 || (depth % 64) != 0) {
+  if (depth < 64 || ((depth / group) % 64) != 0) {
```
Comment (nadavrot) [inline lib/Backends/CPU/Transforms.cpp]: @beicy Can you extract this expression into a new line and add a comment and perhaps a variable names? 

// Make sure that the depth group is divisible by 64. 
bool divisibleBy64 = ((depth / group) % 64) != 0;

File: lib/Backends/CPU/libjit/libjit_conv.cpp [hunk 4]
```diff
@@ -234,9 +236,18 @@ void libjit_convDKKC8_f(float *outW, const float *inW, const float *filterW,
...
                         const size_t *biasWdims, size_t filterSize,
-                        size_t stride, size_t pad, unsigned pixelScanFirst,
-                        unsigned numDepthRegs, unsigned sizeGroupY,
-                        unsigned depthStrips) {
+                        size_t stride, size_t pad, size_t group,
+                        unsigned pixelScanFirst, unsigned numDepthRegs,
+                        unsigned sizeGroupY, unsigned depthStrips) {
+  size_t inChannels = inWdims[3];
+  size_t outChannels = outWdims[3];
+  assert(inChannels % group == 0 &&
```
Comment (nadavrot) [inline lib/Backends/CPU/libjit/libjit_conv.cpp]: Please don't use asserts in libjit. The code is compiled in debug mode and the asserts either remain in the code or are not effective, depending on how we set the optimization pipeline.

File: tests/unittests/JITTest.cpp (hunk 2 above)
Comment (artemrakhov) [reply tests/unittests/JITTest.cpp]: Can we use `ExecutionEngine::run({}, {})` with both Interpreter and CPU backends? So `visibilityKind` is private for filter, but public for final results?

File: lib/Backends/CPU/libjit/libjit_conv.cpp [hunk 5]
```diff
@@ -249,16 +260,21 @@ void libjit_convDKKC8_f(float *outW, const float *inW, const float *filterW,
     // Later we will accumulate values into this slice.
     libjit_conv_init_output_with_bias(n, outW, biasW, outWdims, biasWdims);
 
-    // For each output channel, process [numDepthRegs x float8] elements.
-    for (size_t d = 0; d < outWdims[3]; d += 8 * numDepthRegs * depthStrips) {
+    // For each group of input channels:
+    for (size_t g = 0; g < group; g++) {
+
+      // For each output channel, process [numDepthRegs x float8] elements.
+      for (size_t d = g * outCperG; d < (g + 1) * outCperG;
```
Comment (nadavrot) [inline lib/Backends/CPU/libjit/libjit_conv.cpp]: Please extract the complex expressions and give them a name:

// Calculate the  .... 
size_t startChannelIndex =  g * outCperG;
size_t endChannelIndex =  (g + 1) * outCperG;

File: lib/Backends/CPU/Transforms.cpp (hunk 3 above)
Comment (artemrakhov) [inline lib/Backends/CPU/Transforms.cpp]: Why not `(depth / group) < 64`?

Comment (nadavrot) [reply lib/Backends/CPU/Transforms.cpp]: @artemrakhov The code processes the weight in chunks of 64 elements.

Comment (artemrakhov) [reply lib/Backends/CPU/Transforms.cpp]: I understand. 

We know that depth is greater than 0 and is divisible by group.
Condition `((depth / group) % 64) != 0` (or `(depth % 64) != 0`) includes `(depth / group) < 64` (or `group < 64`). Because the smallest number divisible by 64 is 64.

In other words, my point is that the whole line could if replaced with:
```
  // Make sure that the depth group is divisible by 64.
  bool divisibleBy64 = ((depth / group) % 64) == 0;
  if (!divisibleBy64) {
```

Comment (nadavrot) [reply lib/Backends/CPU/Transforms.cpp]: Oh yes!  In that case, we can just write:

  // Make sure that the depth group is divisible by 64.
  if (((depth / group) % 64) != 0) {

File: tests/unittests/JITTest.cpp (hunk 2 above)
Comment (artemrakhov) [reply tests/unittests/JITTest.cpp]: @nadavrot I think the idea here was to test DKKC8 path (when `optimizeCPUConv()` triggers). We need >=64 channels to trigger it. Maybe it's a bit fragile, because it depends on `64` being hardcoded over there. Would it be better to test `libjit_conv.cpp`'s functions directly, instead of going through Graph and EE ?

Comment (nadavrot) [reply tests/unittests/JITTest.cpp]: @artemrakhov We are going to have hundreds of different optimizations in different backends. I don't think that it makes sense to have a CPU-only test for this specific operator. Let's add a new JIT test like the other ones that we already have that checks the new grouped convolution. This is how we test the regular convolution.

Comment (beicy) [reply tests/unittests/JITTest.cpp]: @artemrakhov Yes, we can use ExecutionEngine::run({}, {}) for both backends,  as long as the var is not loaded from tensor.

File: lib/Backends/CPU/libjit/libjit_conv.cpp (hunk 4 above)
Comment (beicy) [reply lib/Backends/CPU/libjit/libjit_conv.cpp]: Got it. Btw, there are asserts in other functions of this file. Will submit another PR to clean up the assert.

File: lib/Backends/CPU/LLVMIRGen.cpp [hunk 6]
```diff
@@ -1459,10 +1459,15 @@ void LLVMIRGen::generateLLVMIRForInstr(llvm::IRBuilder<> &builder,
     auto *kernel = emitConstSizeT(builder, CI->getKernel());
     auto *stride = emitConstSizeT(builder, CI->getStride());
     auto *pad = emitConstSizeT(builder, CI->getPad());
+    auto *group = emitConstSizeT(builder, CI->getGroup());
 
     size_t inChannels = src->dims()[3];
     size_t outChannels = src->dims()[3];
 
+    assert(inChannels % CI->getGroup() == 0 &&
```
Comment (rdzhabarov) [inline lib/Backends/CPU/LLVMIRGen.cpp]: this is probably redundant, can you add an instruction verifier for that instead, similar to CPUMaxSplat.

File: tests/unittests/BackendTestUtils.cpp [hunk 7]
```diff
@@ -432,6 +432,39 @@ void inferSmallConv(Tensor *inputs, Tensor *out, BackendKind kind) {
   out->copyFrom(&result->getVariable()->getPayload());
 }
 
+void inferGroupConv(Tensor *out, BackendKind kind) {
+  ExecutionEngine EE_(kind);
+  auto &mod_ = EE_.getModule();
+  auto *F_ = mod_.createFunction("main");
+
+  auto *input = mod_.createVariable(ElemKind::FloatTy, {1, 2, 1, 32}, "inp\
```
Comment (rdzhabarov) [inline tests/unittests/BackendTestUtils.cpp]: var name looks weird :)

File: tests/unittests/JITTest.cpp [hunk 8]
```diff
@@ -561,6 +561,20 @@ TEST_P(BackendCorrectnessTest, smallConv) {
...
+/// This test targets the DKKC8 optimization.
+TEST_P(CPUOnly, groupConvTest) {
+  std::array<size_t, 4> S{{1, 2, 1, 128}};
+  llvm::ArrayRef<size_t> shape(S);
+  Tensor out1(ElemKind::FloatTy, shape);
+  Tensor out2(ElemKind::FloatTy, shape);
+  inferGroupConv(&out1, backendKind_);
+  inferGroupConv(&out2, BackendKind::Interpreter);
+
+  for (size_t i = 0; i < 256; i++) {
```
Comment (rdzhabarov) [inline tests/unittests/JITTest.cpp]: any reason not to EXPECT_TRUE(out1.isEqual(out2)); ?

File: lib/Backends/CPU/libjit/libjit_conv.cpp (hunk 4 above)
Comment (rdzhabarov) [reply lib/Backends/CPU/libjit/libjit_conv.cpp]: see also my comment about validating these params in instr verify.

File: tests/unittests/JITTest.cpp (hunk 8 above)
Comment (beicy) [reply tests/unittests/JITTest.cpp]: I used isEqual. For release mode, it works. For debug_mode, it reported false. Tried to output all the elements for debug_mode, but it looks to me the values are equal. Just put  this version here first and still checking isEqual.

Comment (artemrakhov) [reply tests/unittests/JITTest.cpp]: This means that there's very small difference between numbers, e.g. digit number 6 after floating point. Release and debug mode are expected to be a bit different sometimes, because of fast-math optimizations etc.

Perhaps need to adjust allowed error for `isEqual()`.

Comment (rdzhabarov) [reply tests/unittests/JITTest.cpp]: Yup, as Artem said you can tune, float allowedError = 0.0001.

**Comment by nadavrot** [comment]
LGTM!

File: tests/unittests/BackendTestUtils.cpp [hunk 9]
```diff
@@ -432,6 +432,39 @@ void inferSmallConv(Tensor *inputs, Tensor *out, BackendKind kind) {
   out->copyFrom(&result->getVariable()->getPayload());
 }
 
+void inferGroupConv(Tensor *out, BackendKind kind) {
+  ExecutionEngine EE(kind);
+  auto &mod = EE.getModule();
+  auto *F = mod.createFunction("main");
+
+  auto *input = mod.createVariable(ElemKind::FloatTy, {1, 2, 1, 32}, "inp\
```
Comment (artemrakhov) [inline tests/unittests/BackendTestUtils.cpp]: I'm a bit confused about "inp\
Could you name it "input", and move the whole word to the next line? Or have clang-format move it for you.
